### ISSUE TITLE
Table structure issue catalog_product_entity (sku field is allowed nulls)

### ISSUE THREAD

**Issue by stephan-cream** [magento/magento2#27411]
### Preconditions (*)
1. Any Magento version (tested on all significant releases, 2.1, 2.2, 2.3)
2. Reproduced on 2.4-develop

### Steps to reproduce (*)
1. Add a product either through migration, a custom importer or directly in the database.
2. Leave out the field SKU (in other words, leave it NULL).
3. Go to the product edit page in the backoffice.

### Expected result (*)
1. Should not be possible to add the product OR should not throw an error in the backoffice.

### Actual result (*)
1. Product is added without issues.
2. When attempting to edit the product in the backoffice the following error is thrown:
`Fatal error: Uncaught TypeError: Argument 1 passed to Magento\Catalog\Model\ProductRepository::prepareSku() must be of the type string, null given, called in /vendor/magento/module-catalog/Model/ProductRepository.php on line 790 and defined in /vendor/magento/module-catalog/Model/ProductRepository.php:800 Stack trace: 
#0  /vendor/magento/module-catalog/Model/ProductRepository.php(790): Magento\Catalog\Model\ProductRepository->prepareSku(NULL) 
#1 /vendor/magento/module-catalog/Model/ProductRepository.php(349): Magento\Catalog\Model\ProductRepository->saveProductInLocalCache(Object(Magento\Catalog\Model\Product\Interceptor), '3e07f167255019f...') 
#2 /vendor/magento/module-catalog/Model/ProductRepository.php(314): Magento\Catalog\Model\ProductRepository->cacheProduct('3e07f167255019f...', Object(Magento\Catalog\Model\Product in /vendor/magento/module-catalog/Model/ProductRepository.php on line 800`

**Comment by stephan-cream** [comment]
I believe this to be a mistake in the structure design, as SKU is required throughout Magento.

**Comment by rodrigowebjump** [comment]
@magento give me 2.4-develop instance

**Comment by magento-engcom-team** [comment]
Hi @rodrigowebjump. Thank you for your request. I'm working on Magento 2.4-develop instance for you

**Comment by magento-engcom-team** [comment]
Hi @rodrigowebjump, here is your [Magento instance](https://i-27411-2-4-develop.instances.magento-community.engineering).
Admin access: [https://i-27411-2-4-develop.instances.magento-community.engineering/admin_399e](https://i-27411-2-4-develop.instances.magento-community.engineering/admin_399e)
Login: `7deac959` Password: `acb0d5600985`
Instance will be terminated in up to 3 hours.

**Comment by stephan-cream** [comment]
This is a database issue, unless you can directly insert the product in the database (or use custom code to import it) without an SKU this is not 'testable' on an Magento github generated development instance.

**Comment by pmourer** [comment]
There are likely countless places in core code where it is not expected that a valid product could have a `NULL` sku. This needs to be addressed.
```
PHP Fatal error:  Uncaught TypeError: Argument 1 passed to Magento\Catalog\Model\ProductRepository::getProductFromLocalCache() must be of the type string, null given, called in /vendor/magento/module-catalog/Model/ProductRepository.php on line 270 and defined in /vendor/magento/module-catalog/Model/ProductRepository.php:751
Stack trace:
#0 /vendor/magento/module-catalog/Model/ProductRepository.php(270): Magento\Catalog\Model\ProductRepository->getProductFromLocalCache(NULL, 'adf0bb68589672e...')
#1 /generated/code/Magento/Catalog/Model/ProductRepository/Interceptor.php(24): Magento\Catalog\Model\ProductRepository->get(NULL, false, NULL, false)
#2 /vendor/magento/module-bundle/Model/LinkManagement.php(253): Magento\Catalog\Model\ProductRepository\Interceptor->get(NULL)
#3 /vendor/magento/module-bundle/Model/Option/SaveAction.php(113): Magento\Bundle\Model\LinkManagement->addChild(Object(Magen in /vendor/magento/module-catalog/Model/ProductRepository.php on line 751
```

When loading admin grid for products I also got an error, but cannot find now as i cleared all products with null skus from my database.

**Comment by pmourer** [comment]
`vendor/magento/module-catalog/etc/db_schema.xml` line 16
```
        <column xsi:type="varchar" name="sku" nullable="true" length="64" comment="SKU"/>
```
`vendor/magento/module-catalog/Api/Data/ProductInterface.php` lines 74-79

```
    /**
     * Product sku
     *
     * @return string
     */
    public function getSku();
```

one of these is not like the other ones.

**Comment by Pavel-Fekete** [comment]
Reproducible on the latest 2.4-develop
Product is added without issues,  even if  SKU value is '_NULL_'

![database](https://user-images.githubusercontent.com/67371013/102361991-8607ed00-3fbc-11eb-9dfc-ec6485a2a2b1.png)

**Comment by magento-engcom-team** [comment]
:white_check_mark: Confirmed by @engcom-Bravo 
 Thank you for verifying the issue. Based on the provided information internal tickets `MC-39963` were created

**Issue Available**: @engcom-Bravo, _You will be automatically unassigned. Contributors/Maintainers can claim this issue to continue. To reclaim and continue work, reassign the ticket to yourself._

**Comment by magento-engcom-team** [comment]
Unfortunately, we are archiving this ticket now as it did not get much attention from both Magento Community and Core developers for an extended period. This is done in an effort to create a quality, community-driven backlog which will allow us to allocate the required attention more easily.

Please feel free to comment or reopen according to the [Issue reporting guidelines](https://github.com/magento/magento2/wiki/Issue-reporting-guidelines) 
the ticket if you are still facing this issue on the latest `2.x-develop` branch. Thank you for collaboration.

**Comment by dverkade** [comment]
@magento-engcom-team  please reopen ticket as this is reproduced on Magento 2.4.X and should be fixed.

**Comment by gabrieldagama** [comment]
Hi @stephan-cream. Thank you for your report. 
The issue has been fixed in magento/magento2#32532 by @ihor-sviziev in 2.4-develop branch
Related commit(s):
 - [a369d5a510f7bbbdf790bfdeffcc5fc9126949e1](https://github.com/magento/magento2/commit/a369d5a510f7bbbdf790bfdeffcc5fc9126949e1)

The fix will be available with the upcoming 2.4.3 release.
